File: Seafood Guide/Lingo/ViewControllers/LingoTableViewController.m
//
//  LingoTableViewController.m
//  Seafood Guide
//
//  Created by Jon Brown on 9/1/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

#import "LingoTableViewController.h"
#import "CoreDataTableViewController.h"
#import "Lingo.h"
#import "AppDelegate.h"
#import "DetailLingoViewController.h"

@implementation LingoTableViewController
@synthesize window = _window;
@synthesize fetchedResultsController = __fetchedResultsController;
@synthesize managedObjectContext = __managedObjectContext;


- (NSManagedObjectContext *)managedObjectContext
{
    if (!__managedObjectContext)
    {
        __managedObjectContext = [(AppDelegate *)[[UIApplication sharedApplication] delegate] managedObjectContext];
    }
    return __managedObjectContext;
    
}


- (void)setupFetchedResultsController

{
    
    // 1 - Decide what Entity you want
    NSString *entityName = @"Lingo"; // Put your entity name here
    //NSLog(@"Setting up a Fetched Results Controller for the Entity named %@", entityName);
    
    // 2 - Request that Entity
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:entityName];
    
    
    // 3 - Filter it if you want
    
    //predicateKey = @"";
    
    // 4 - Sort it if you want
    request.sortDescriptors = [NSArray arrayWithObject:[NSSortDescriptor sortDescriptorWithKey:@"titlenews"
                                                                                     ascending:YES
                                                                                      selector:@selector(localizedCaseInsensitiveCompare:)]];
    
    // 5 - Fetch it
    self.fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:request
                                                                        managedObjectContext:self.managedObjectContext
                                                                          sectionNameKeyPath:@"uppercaseFirstLetterOfName"
                                                                                   cacheName:nil];
    
    
    [self.fetchedResultsController performFetch:nil];
    
    
    
 
}



- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [[self.fetchedResultsController sections] count];
}


- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
    return [sectionInfo name];
}

- (NSInteger)tableView:(UITableView *)table numberOfRowsInSection:(NSInteger)section {
    id <NSFetchedResultsSectionInfo> sectionInfo = [[self.fetchedResultsController sections] objectAtIndex:section];
    return [sectionInfo numberOfObjects];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self setupFetchedResultsController];
    
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    static NSString *CellIdentifier = @"lingocells";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    
    
    // Configure the cell...
    
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    
    Lingo *lingo = [self.fetchedResultsController objectAtIndexPath:indexPath];
    
    cell.textLabel.text = lingo.titlenews;
    
    //NSLog(@"%@", lingo.titlenews);
    
    return cell;
    
}

- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView
{
    return [[UILocalizedIndexedCollation currentCollation] sectionIndexTitles];
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index {
    return [[UILocalizedIndexedCollation currentCollation] sectionForSectionIndexTitleAtIndex:index];
}

- (id)initWithLingoSize:(NSString *)size
{
    self = [super initWithStyle:UITableViewStylePlain];
    
    return self;
}

- (id)initWithStyle:(UITableViewStyle)style
{
    return [self initWithLingoSize:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationItem.backBarButtonItem.title = @"Back";
}

- (void)viewDidUnload
{
    [super viewDidUnload];

}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

#pragma mark - Table view sends data to detail view
// Core data to detail view

-(void)tableView:(UITableView *)aTableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    DetailLingoViewController *detailViewController = [[DetailLingoViewController alloc]init];
    
    NSInteger sumSections = 0;
    for (int i = 0; i < indexPath.section; i++) {
        long rowsInSection = [self.tableView numberOfRowsInSection:i];
        sumSections += rowsInSection;
    }
    NSInteger currentRow = sumSections + indexPath.row;
    
   
    Lingo *allLingo = [[self.fetchedResultsController fetchedObjects] objectAtIndex:currentRow];
    [detailViewController setItem:allLingo];

    [self.navigationController pushViewController:detailViewController
                                         animated:YES];
    
    self.navigationItem.backBarButtonItem.title = @"";
}
-(void)viewDidAppear:(BOOL)animated
{
    [self.navigationItem setHidesBackButton:NO];
}

@end

File: Seafood Guide/Lingo/ViewControllers/LingoTableViewController.swift
//  Converted to Swift 5.1 by Swiftify v5.1.26565 - https://objectivec2swift.com/
//
//  LingoTableViewController.swift
//  Seafood Guide
//
//  Created by Jon Brown on 9/1/14.
//  Copyright (c) 2014 Jon Brown Designs. All rights reserved.
//

import UIKit
import CoreData

class LingoTableViewController: UITableViewController, UINavigationBarDelegate, UINavigationControllerDelegate {
    var uppercaseFirstLetterOfName = ""
    var predicateKey = ""

    var window: UIWindow?
    var fetchedResultsController: NSFetchedResultsController<NSFetchRequestResult>?

    lazy var managedObjectContext: NSManagedObjectContext? = {
        return (UIApplication.shared.delegate as? AppDelegate)?.managedObjectContext
    }()
    
    var sectionIndexTitles: [AnyHashable] = []

    init(lingoSize size: String?) {
        super.init(style: .plain)
    }

    func setupFetchedResultsController() {

        let entityName = "Lingo" // Put your entity name here

        let request = NSFetchRequest<NSFetchRequestResult>(entityName: entityName)
        
        request.sortDescriptors = [NSSortDescriptor.init(key: "titlenews", ascending: true)]

        if let managedObjectContext = managedObjectContext {
            fetchedResultsController = NSFetchedResultsController(fetchRequest: request, managedObjectContext: managedObjectContext, sectionNameKeyPath: "uppercaseFirstLetterOfName", cacheName: nil)
        }

        do {
            try fetchedResultsController?.performFetch()
        } catch {
        }

    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return fetchedResultsController?.sections?.count ?? 0
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        weak var sectionInfo = fetchedResultsController?.sections?[section]
        return sectionInfo?.name
    }

    override func tableView(_ table: UITableView, numberOfRowsInSection section: Int) -> Int {
        weak var sectionInfo = fetchedResultsController?.sections?[section]
        return sectionInfo?.numberOfObjects ?? 0
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        setupFetchedResultsController()

    }

    static let tableViewCellIdentifier = "lingocells"

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        var cell = tableView.dequeueReusableCell(withIdentifier: LingoTableViewController.tableViewCellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: LingoTableViewController.tableViewCellIdentifier)
        }
        cell?.accessoryType = .disclosureIndicator

        let lingo = fetchedResultsController?.object(at: indexPath) as? Lingo

        cell?.textLabel?.text = lingo?.titlenews

        return cell!

    }

    override func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return UILocalizedIndexedCollation.current().sectionIndexTitles
    }

    override func tableView(_ tableView: UITableView, sectionForSectionIndexTitle title: String, at index: Int) -> Int {
        return UILocalizedIndexedCollation.current().section(forSectionIndexTitle: index)
    }

    convenience override init(style: UITableView.Style) {
        self.init(lingoSize: nil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationController?.navigationBar.tintColor = UIColor(named: "Text Color")!
        navigationItem.backBarButtonItem?.title = "Back"
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

// MARK: - Table view sends data to detail view
    override func tableView(_ aTableView: UITableView, didSelectRowAt indexPath: IndexPath) {

        let detailViewController = DetailLingoViewController.init(style: .grouped)
        detailViewController.tableView.backgroundColor = UIColor(named: "Background")
        
        var sumSections = 0
        for i in 0..<indexPath.section {
            let rowsInSection = tableView.numberOfRows(inSection: i)
            sumSections += rowsInSection
        }
        let currentRow = sumSections + indexPath.row


        let allLingo = fetchedResultsController?.fetchedObjects?[currentRow] as? Lingo
        detailViewController.item = allLingo

        navigationController?.pushViewController(detailViewController, animated: true)

        navigationItem.backBarButtonItem = UIBarButtonItem(title: "Back", style: .plain, target: nil, action: nil)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        navigationItem.hidesBackButton = false
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}
